otherlibs/unix/socketaddr.c

```diff
@@ -69,7 +69,22 @@ void caml_unix_get_sockaddr(value vaddr,
         caml_unix_error(ENOENT, "", path);
       }
       memmove(s_unix->sun_path, String_val(path), len + 1);
-      *addr_len = offsetof(struct sockaddr_un, sun_path) + len;
+      /* Unnamed socket */
+      if (len == 0)  {
+        /* sizeof(sun_family) on linux,  to trigger autobind according to
+           https://man7.org/linux/man-pages/man7/unix.7.html
+           sizeof(sun_family) + sizeof(sun_len) on Haiku */
+        *addr_len = offsetof(struct sockaddr_un, sun_path);
+      } else {
+        if (Byte(path, 0) == 0) {
+          /* Abstract socket */
+          *addr_len = offsetof(struct sockaddr_un, sun_path) + len;
+        } else {
+          /* Regular name socket, count the trailing 0 according to
+            https://man7.org/linux/man-pages/man7/unix.7.html */
+          *addr_len = offsetof(struct sockaddr_un, sun_path) + len + 1;
+        }
+      }
```

Comment:
Is this cursed?

```suggestion
      *addr_len = offsetof(struct sockaddr_un, sun_path)
                  + (len == 0 ? 0 : !!Byte(path, 0));
```

Reply:
I still think that without code-golfing too hard a condensed version can also be readable. I understand it's a matter of personal style but the one above feels too long.

```c
int need_trailing_null = len != 0 && Byte(path, 0) != 0;
*addr_len = offsetof(struct sockaddr_un, sun_path) + len + need_trailing_null;
```